cpp/src/arrow/array/util.cc

```diff
@@ -502,6 +502,11 @@ class NullArrayFactory {
     return Status::OK();
   }
 
+  Status Visit(const ExtensionType& type) {
+    RETURN_NOT_OK(VisitTypeInline(*type.storage_type(), this));
+    return Status::OK();
+  }
+
```

Comment:
Should we also add a C++ test here? https://github.com/apache/arrow/blob/87e2ad5b2f2cdaf1e469b1cda1a2899a747464b6/cpp/src/arrow/array/array_test.cc#L334

Reply:
+1 for adding a test

Reply:
Added the `smallint()` extension type to that test.